### ISSUE TITLE
Error "Failed to build gem native extension." when running "gem install openssl" in Fedora

### ISSUE THREAD

**Issue by sergioro9** [ruby/openssl#716]
When I try to install `openssl` gem in Fedora 38 server I get this error:

```
gem install openssl
Building native extensions. This could take a while...
ERROR:  Error installing openssl:
	ERROR: Failed to build gem native extension.

    current directory: /home/deploy/.rvm/gems/ruby-3.0.3/gems/openssl-3.2.0/ext/openssl
/home/deploy/.rvm/rubies/ruby-3.0.3/bin/ruby -I /home/deploy/.rvm/rubies/ruby-3.0.3/lib/ruby/3.0.0 -r ./siteconf20240205-1375528-r7q9az.rb extconf.rb
checking for rb_io_descriptor()... no
checking for rb_io_maybe_wait(0, Qnil, Qnil, Qnil) in ruby/io.h... no
checking for t_open() in -lnsl... no
checking for socket() in -lsocket... no
checking for openssl/ssl.h... yes
checking for CRYPTO_malloc() in -lcrypto... yes
checking for SSL_new() in -lssl... yes
checking for LIBRESSL_VERSION_NUMBER in openssl/opensslv.h... no
checking for OpenSSL version >= 1.0.2... no
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/home/deploy/.rvm/rubies/ruby-3.0.3/bin/$(RUBY_BASE_NAME)
	--with-openssl-dir
	--with-openssl-include
	--without-openssl-include=${openssl-dir}/include
	--with-openssl-lib
	--without-openssl-lib=${openssl-dir}/lib
	--with-kerberos-dir
	--without-kerberos-dir
	--with-kerberos-include
	--without-kerberos-include=${kerberos-dir}/include
	--with-kerberos-lib
	--without-kerberos-lib=${kerberos-dir}/lib
	--with-nsl-dir
	--without-nsl-dir
	--with-nsl-include
	--without-nsl-include=${nsl-dir}/include
	--with-nsl-lib
	--without-nsl-lib=${nsl-dir}/lib
	--with-nsllib
	--without-nsllib
	--with-socket-dir
	--without-socket-dir
	--with-socket-include
	--without-socket-include=${socket-dir}/include
	--with-socket-lib
	--without-socket-lib=${socket-dir}/lib
	--with-socketlib
	--without-socketlib
	--with-crypto-dir
	--without-crypto-dir
	--with-crypto-include
	--without-crypto-include=${crypto-dir}/include
	--with-crypto-lib
	--without-crypto-lib=${crypto-dir}/lib
	--with-cryptolib
	--without-cryptolib
	--with-ssl-dir
	--without-ssl-dir
	--with-ssl-include
	--without-ssl-include=${ssl-dir}/include
	--with-ssl-lib
	--without-ssl-lib=${ssl-dir}/lib
	--with-ssllib
	--without-ssllib
extconf.rb:136:in `<main>': OpenSSL >= 1.0.2 or LibreSSL >= 3.1.0 is required (RuntimeError)

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/deploy/.rvm/gems/ruby-3.0.3/extensions/x86_64-linux/3.0.0/openssl-3.2.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /home/deploy/.rvm/gems/ruby-3.0.3/gems/openssl-3.2.0 for inspection.
Results logged to /home/deploy/.rvm/gems/ruby-3.0.3/extensions/x86_64-linux/3.0.0/openssl-3.2.0/gem_make.out
```
According to the error message I do not have OpenSSL version >= 1.0.2 installed. But that is not the case, I have OpenSSL 3.0.9 and `openssl-devel` library:

```
$ openssl version
OpenSSL 3.0.9 30 May 2023 (Library: OpenSSL 3.0.9 30 May 2023)

$ sudo dnf install openssl-devel
Package openssl-devel-1:3.0.9-2.fc38.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```

The ruby version is 3.0.3 which I installed with the command:

```
rvm install 3.0.3 --with-openssl-dir=$HOME/.rvm/usr
```

Could someone help me fix this issue? In particular, why `gem install openssl` says that I don't have OpenSSL >= 1.0.2 installed when I do have it.

**Comment by sergioro9** [comment]
After looking into the issue `https://github.com/ruby/openssl/issues/686` I found out that the header files of my rvm ruby installation are `OpenSSL 1.0.1` which is different from my system-wide OpenSSL version.

```
$ grep OPENSSL_VERSION /home/deploy/.rvm/usr/include/openssl/opensslv.h
#define OPENSSL_VERSION_NUMBER	0x1000109fL
#define OPENSSL_VERSION_TEXT	"OpenSSL 1.0.1i-fips 6 Aug 2014"
#define OPENSSL_VERSION_TEXT	"OpenSSL 1.0.1i 6 Aug 2014"
#define OPENSSL_VERSION_PTEXT	" part of " OPENSSL_VERSION_TEXT

$ grep OPENSSL_VERSION /usr/include/openssl/opensslv.h
# define OPENSSL_VERSION_MAJOR  3
# define OPENSSL_VERSION_MINOR  0
# define OPENSSL_VERSION_PATCH  9
...
```

It seems that I need to reinstall ruby with something like `rvm reinstall 3.0.3 --with-openssl-dir=OpenSSL3-0-9-Directory`. But I am not sure what the value for `--with-openssl-dir` should be.

**Comment by rhenium** [comment]
Installing `openssl` gem by default searches the same paths used when your Ruby was installed. Your Ruby 3.0.3 seems to be using the OpenSSL 1.0.1i (which is a very, very old release with known vulnerabilities - I'd upgrade it now), which is no longer supported by the latest `openssl` gem release, hence the first error.

Fedora's OpenSSL 3.0.9 is in `/usr`. Does `gem install openssl -- --with-openssl-dir=/usr` work?

**Comment by junaruga** [comment]
In your case, I assume that you installed Ruby 3.0.3 via RVM when the `openssl-devel` RPM veresion was 1.0.1i in the past, then you upgraded Fedora to new versions, and now the `openssl-devel` RPM version is 3.0.9.

**Comment by sergioro9** [comment]
I solved the issue by installing Ruby 3.2.3 which supports OpenSSL 3, which is the OpenSSL version installed in my Fedora 38 server.

I used this command to install ruby:

```
rvm install 3.2.3
```
